### PR TITLE
feat(docker): reduce the size of alpine docker images

### PR DESCRIPTION
**Pull request by netroy**
* reduce redundant code between the two alpine Dockerfiles
* reduce the docker image build times
* reduce the `n8n` image size from 671MB to 460MB
* reduce the `n8n-custom` image size from 1.17GB to 462MB
* reduce the size of npm tree by making all typing packages devDependencies

Partially resolves this ticket: N8N-3252

### CODE REVIEW

File: docker/images/n8n-custom/Dockerfile [hunk 1]
```diff
@@ -1,55 +1,34 @@
...
-
-# 2. Start with a new clean image with just the code that is needed to run n8n
-FROM node:16-alpine
 
-USER root
+USER node
 
-RUN apk add --update graphicsmagick tzdata tini su-exec git
+RUN \
+	npm ci && \
```
Comment (agentreno) [inline docker/images/n8n-custom/Dockerfile:19]: any value in running `npm ci` separately to cache the dependencies in a layer of their own for the times we only change source code? I'm assuming `npm run build` generates changes on every source code change.

**Review by agentreno — COMMENTED** [review]
looks good! :+1: i left a few comments (some for my own education, one to discuss)

File: .github/workflows/docker-base-image.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      - name: Build
+        uses: docker/build-push-action@v2
+        with:
+          context: ./docker/images/n8n-base
+          build-args: |
+            NODE_VERSION=${{github.event.inputs.node_version}}
+          platforms: linux/amd64,linux/arm64,linux/arm/v7
+          push: true
+          tags: |
+            ${{ secrets.DOCKER_USERNAME }}/base:${{ github.event.inputs.node_version }}
```
[the remark this answers is not in the document]
Comment (netroy) [reply .github/workflows/docker-base-image.yml:35]: this the nodejs version. for our own version we use `N8N_VERSION`

File: docker/images/n8n-custom/Dockerfile (hunk 1 above)
Comment (netroy) [reply docker/images/n8n-custom/Dockerfile:19]: The issue here is that `npm ci` is installing dev-dependencies as well, which are needed for the `npm run build`. These are later deleted by the `npm prune`. 
If we split this up, we end up with a significantly larger image, as the layer with `npm ci` contains all the dev-dependencies, despite us not needing them after the build.

I plan to optimize these layers in the future to reduce the image sizes even further, but none of those plans fit into the scope of this PR IMO.

Comment (agentreno) [reply docker/images/n8n-custom/Dockerfile:19]: makes sense :+1: thanks

File: .github/workflows/docker-base-image.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (netroy) [reply .github/workflows/docker-base-image.yml:35]: this workflow will be manually triggered only when the image needs to be updated. and the version of the image will match the version of the nodejs image we are pulling from docker hub.

File: .github/workflows/docker-base-image.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        required: true
+        default: '16'
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v1
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
```
[the remark this answers is not in the document]
Comment (netroy) [reply .github/workflows/docker-base-image.yml:18]: From what I understand we need qemu to be able to build some of the multi-arch images.

**Review by csuermann — DISMISSED** [review]

**Review by csuermann — APPROVED** [review]

**Comment by janober** [comment]
Got released with `n8n@0.193.0`

**Pull request merged**